### ISSUE TITLE
WebSocket terminated during authentication (magic code) after 403 returned from GetSessionId call

### ISSUE THREAD

**Issue by darrenj** [Microsoft/BotFramework-WebChat#1097]
When using the CDN hosted webchat (0.14.0) and a Bot configured to use the new authentication support your unable to complete authentication involving magic codes due to the WebChat control terminating the WebSocket.

Hosting the WebChat control yourself, providing an OriginUrl and exchanging the secret for a token means the Magic Code step can be avoided however we should be able to complete authentication with magic codes using the standard WebChat control otherwise this blocks an important set of scenarios working in normal WebChat mode - especially during development.

The Web page uses the WebChat control like this:

`<script src="https://cdn.botframework.com/botframework-webchat/0.14.0/botchat.js"></script>`

…

  ```
BotChat.App({
            directLine: { secret: 'YOURSECRETHERE' },
            user: { id: 'darren' },
            bot: { id: 'YOURBOTID' },
            resize: 'detect'
        }, document.getElementById("bot"));
```
When clicking the Signin button in the OAuthCard you get a popup and complete authentication and are given a magic number. However as a result of this the WebSocket has been terminated on the call to /GetSessionId, we think this is because the secret doesn't contain the trusted Origin Urls which in this case is expected (we are using magic codes).

![image](https://user-images.githubusercontent.com/33519520/44254182-fdb18b80-a1f9-11e8-8a51-f98b48cc04c3.png)

Expected behaviour is for WebChat to show a prompt - e.g. "please enter the magic code to complete authentication" enabling the user to copy/paste the magic code and authentication completes.

Repro is here: https://aadv2botdarrenj.azurewebsites.net/default.htm

Type Calendar and then click the sign-in button.

**Comment by compulim** [comment]
The Web Socket connection is maintained by Direct Line JS over [this line of code](https://github.com/Microsoft/BotFramework-DirectLineJS/blob/master/src/directLine.ts#L644).

It seems when `getSessionId()` failed ([over here](https://github.com/Microsoft/BotFramework-DirectLineJS/blob/master/src/directLine.ts#L502)), we send it to [`catchPostError`](https://github.com/Microsoft/BotFramework-DirectLineJS/blob/master/src/directLine.ts#L577) for error handling. And if the error is 403, we say the [token is expired and kill the connection](https://github.com/Microsoft/BotFramework-DirectLineJS/blob/master/src/directLine.ts#L400).

Two questions:

- Why we get 403 when the user completed the magic code flow to `/v3/directline/session/getsessionid`?
- Why should we handle 403 from `/getsessionid` the same way as other calls like `postActivity`?

**Comment by mingweiw** [comment]
The 403 in this case isn't because the token expired. The getsessionid call has additional validation requirements on the token. We probably should handle it differently. I can add a specific error code to this 403 on the service side and on the client you use that to decide.

**Comment by compulim** [comment]
Sounds good. So let me rephrase that in bullet form:

- `/getsessionid` could fail in two cases:
   - OAuth flow is not working
      - Today, server return 403, which confused Direct Line JS
      - Tomorrow, we will fix it by either:
         - Server return 403, plus additional error code in the response body
         - Server return something other than 403
   - Web Chat token is expired
      - Today, server return 403
      - Tomorrow, server will still return 403

Is that correct?

**Comment by mingweiw** [comment]
Correct.

**Comment by stevekay72** [comment]
I'm a little confused here.
We have taken the webchat code in this repo so that we can brand it ourselves and extend for auto login, etc and are facing the same problem.  If I make it use a signin card then all is well - get the magic code and it works.  If we use an OAuthCard then we get the same problem.
Is there a work around or is the issue going to be fixed?

**Comment by mingweiw** [comment]
OAuthCard supports a flow that has no mogic code. Signin just works normally. To get that to work you'll need to acquire a different directline token and initialize webchat with that. If you do that, this problem goes away. However we don't have official documentation for that yet. We are working on making that flow simpler to setup and more secure. We are targeting that and the docs for Ignite in Sep.

There is also a PR in progress that fixes this websocket termination issue. That'll allow the magic code flow to complete.

**Comment by stevekay72** [comment]
Thanks for the explanation @mingweiw .  

So for the time being we can use the secret to generate a token using /v3/directline/tokens/generate and use that in the web chat, but there will be a fix coming in a PR that will mean that it is not necessary - correct?

Which PR is it so I can watch for it?

**Comment by mingweiw** [comment]
No. :)
1. To get a token that'll allow the getsessionid call to succeed you need to pass an additional parameter in the body. How to do that is not currently documented. We are also making updates there that'll ship soon. When the getssesionid call succeeds, you'll also see the magic code step disappear.
2. The PR will fix the websocket termination issue. The getssesionid call will still fail, you'll still get the magic code. But you can continue the chat.

**Comment by compulim** [comment]
403 termination issue is fixed in #1131.
